File: UniqueBinarySearchTreesII.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        vector<TreeNode*> trees;
        if(n==0) return trees;
        trees=generateTrees(1, n);
        return trees;
    }
    
    vector<TreeNode*> generateTrees(int begin, int end){
        vector<TreeNode*> trees;
        if(begin>end){
            trees.push_back(NULL);
            return trees;
        }
        for(int i=begin;i<=end;i++){
            vector<TreeNode*> left_trees;
            vector<TreeNode*> right_trees;
            
            left_trees=generateTrees(begin, i-1);
            right_trees=generateTrees(i+1, end);
            
            for(int j=0;j<left_trees.size();j++){
                for(int k=0;k<right_trees.size();k++){
                    TreeNode *temp=new TreeNode(i);
                    temp->left=left_trees[j];
                    temp->right=right_trees[k];
                    trees.push_back(temp);
                }
            }
        }
        
        return trees;
    }
    
    void printTree(TreeNode *root){
        if(root!=NULL){
            cout<<root->val<<" ";
            printTree(root->left);
            printTree(root->right);
            cout<<endl;
        }
    }
};

int main(){
    int n;
    cin>>n;
    
    Solution *solution=new Solution();
    vector<TreeNode*> result=solution->generateTrees(n);
    for(int i=0;i<result.size();i++){
        solution->printTree(result[i]);
    }
    
    return 0;
    
}

File: UniqueBinarySearchTreesII.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def generateTrees(begin, end):
            trees = []
            if begin>end:
                trees.append(None)
                return trees
            for i in range(begin, end+1):
                left_trees = generateTrees(begin, i-1)
                right_trees = generateTrees(i+1, end)
                
                for l in left_trees:
                    for r in right_trees:
                        temp = TreeNode(i)
                        temp.left = l
                        temp.right = r
                        trees.append(temp)
            return trees
        if n==0:
            return []
        return generateTrees(1, n)
        
    def printTree(self, root):
        if root!=None:
            print(root.val)
            self.printTree(root.left)
            self.printTree(root.right)
            print("\n")
        
if __name__=='__main__':
    n = int(raw_input())
    solution = Solution()
    result = solution.generateTrees(n)
    for r in result:
        solution.printTree(r)
    
